Always insert CEPs into the largest available CES

dlapcevic opened the pull request:
Currently, with FCFS (First Come, First Served) batching mode, CiliumEndpoints are inserted into the first available (non-full, non-empty) CiliumEndpointSlice, which comes in an unordered list from the cache.

The change is to go through the list and pick the largest available CES, instead of the first available. With this change, it’s guaranteed that the creation of CEPs will cause the minimum number of CES updates, using the FCFS batching mode.

Signed-off-by: Dorde Lapcevic <[dordel@google.com](mailto:dordel@google.com)>

alan-kut commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -297,14 +297,11 @@ func (c *cesMgr) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint, ns string)
+		// consistently filled up in order.
+		ces := c.getLargestAvailableCES()
+		if ces != nil {
 			return ces, ces.ces.GetName()
```
Do we need to have a ces.backendMutex locked for this?
It was the case for the previous code:

defer ces.backendMutex.RUnlock()
return ces, ces.ces.GetName()

instead of

ces.backendMutex.RUnlock()
return ces, ces.ces.GetName()

dlapcevic replied:
Added the read lock for it to stay consistent, since the lock is used in other parts of the code when `ces.ces.GetName()` is called.

I think it doesn't need to use defer unlock there, as it doesn't make sense for CES objects to have a changed name.
A different name in the ObjectMeta pretty much just means a different object (CES).
The cesName is used after the lock is released anyway, so it could be changed after as well.

dlapcevic committed:
Always insert CEPs into the largest available CES

Currently, with FCFS (First Come, First Served) batching mode, CiliumEndpoints are inserted into the first available (non-full, non-empty) CiliumEndpointSlice, which comes in an unordered list from the cache.

The change is to go through the list and pick the largest available CES, instead of the first available. With this change, it’s guaranteed that the creation of CEPs will cause the minimum number of CES updates, using the FCFS batching mode.

Signed-off-by: Dorde Lapcevic <dordel@google.com>
